src/misc/misc.h

```diff
@@ -67,8 +67,19 @@ extern float quickselect(float *arr, unsigned int n, unsigned int k);
 
 extern float quickselect_complex(_Complex float *arr, unsigned int n, unsigned int k);
 
-extern void mini_cmdline(int* argcp, char* argv[], int expected_args, const char* usage_str, const char* help_str);
-extern _Bool mini_cmdline_bool(int* argcp, char* argv[], char flag_char, int expected_args, const char* usage_str, const char* help_str);
+extern int mini_cmdline_impl(int* argcp, char* argv[], int expected_args, const char* usage_str, const char* help_str);
+extern int mini_cmdline_bool(int* argcp, char* argv[], char flag_char, int expected_args, const char* usage_str, const char* help_str, _Bool* flag);
+
+#define mini_cmdline(acp, av, e_args, u_str, h_str)			\
+     do {								\
```

Comment:
I would rather avoid these macros and the change in API. How about we 'error' instead of 'exit' and the original versions of 'mini_cmdline' and 'cmdline'. We could then use 'longjmp' excactly as in io_error. What do you think?

Reply:
Since these functions are only used in very few specific locations and are not a central part of the BART framework, I don't think changing the API is all that bad.
Plus, from a design perspective, I don't like functions which job should be only limited with parsing arguments exiting the program on their own. I feel it should be the caller's responsibility to deal with any error and then exit if need be.

The effects on client code should also be fairly minimal, as only custom commands written outside of BART calling `mini_cmdline_bool` directly would be impacted by those changes (and in those cases the code would simply not compile, avoiding running with unexpected behaviour). A comment in the code above the declaration of `mini_cmdline_bool` could then instruct users on how they can fix their code.

Reply:
There is no sensible way to recover from an incorrect command-line. Why do you think it is good from a design perspective to add error handling?

But even worse, you do not even add this error handling, you just re-implement  the same API using macros ;-)

To add insult to injury, it is not even the same API: Instead of calling exit, the macro returns from the calling function. This is really unusual and surprising behavior and now makes it impossible to use these functions anywhere else than in main.

Reply:
I agree with you that recovering from an invalid command line is impossible and that a program should not try to continue afterwards, but instead indicate to the user that an error occurred and terminate. Calling the `exit`function is one way of doing so which works well for standalone programs like BART.

However, in programming languages that have proper exception handling, failure to parse command line arguments is usually handled by throwing an exception that gives the possibility of each caller function to act upon it and do some cleanup if necessary (such as closing file descriptors, freeing heap allocated memory, etc.). Additionally, throughout this process normal stack unwinding takes place to make sure that no memory accidentally becomes unaccounted for. Unfortunately, C has no such mechanism and `longjmp` emulates only part of those functionalities, which is why I tried to avoid relying on it as much as possible.

I also did not know that these functions were intended to be called elsewhere in BART, which make my modifications indeed inappropriate.

Anyway, in light of your comments, I will revert those changes and revert to calling the `error` function which will then make the program jump back to the BART main function (via `longjmp` if enabled) and subsequently return and exit with an error code.

Reply:
Well, there are not necessarily called elsewhere in BART, but there is a lot of software build on top of BART in my lab or reusing parts of BART (and maybe elsewhere). But I tend agree with your analysis, and I agree that the longjmp solution is also not ideal. Maybe the best choice would be to return the error and change all the callers to check it, but this seems overkill at this time.